pkg/volume/csi/csi_plugin.go

```diff
@@ -0,0 +1,251 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package csi
+
+import (
+	"errors"
+	"fmt"
+	"path"
+	"regexp"
+	"time"
+
+	csipb "github.com/container-storage-interface/spec/lib/go/csi"
+	"github.com/golang/glog"
+	api "k8s.io/api/core/v1"
+	meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/kubernetes/pkg/util/mount"
+	kstrings "k8s.io/kubernetes/pkg/util/strings"
+	"k8s.io/kubernetes/pkg/volume"
+)
+
+const (
+	csiName       = "csi"
+	csiPluginName = "kubernetes.io/csi"
+
+	// TODO (vladimirvivien) implement a more dynamic way to discover
+	// the unix domain socket path for each installed csi driver.
+	// TODO (vladimirvivien) would be nice to name socket with a .sock extension
+	// for consistency.
+	csiAddrTemplate = "/var/lib/kubelet/plugins/%v"
+	csiTimeout      = 15 * time.Second
+)
+
+var (
+	// csiVersion supported csi version
+	csiVersion     = &csipb.Version{Major: 0, Minor: 1, Patch: 0}
+	driverNameRexp = regexp.MustCompile(`^[A-Za-z]+(\.?-?_?[A-Za-z0-9-])+$`)
+)
+
+type csiPlugin struct {
+	host volume.VolumeHost
+}
+
+// ProbeVolumePlugins returns implemented plugins
+func ProbeVolumePlugins() []volume.VolumePlugin {
+	p := &csiPlugin{
+		host: nil,
+	}
+	return []volume.VolumePlugin{p}
+}
+
+// volume.VolumePlugin methods
+var _ volume.VolumePlugin = &csiPlugin{}
+
+func (p *csiPlugin) Init(host volume.VolumeHost) error {
+	glog.Info(log("plugin initializing..."))
+	p.host = host
+	return nil
+}
+
+func (p *csiPlugin) GetPluginName() string {
+	return csiPluginName
+}
+
+// GetvolumeName returns a concatenated string of <CSIVolumeSource.Driver~CSIVolumeSource.VolumeHandle>
+// That string value is used in Detach() to extract driver name and volumeName.
+func (p *csiPlugin) GetVolumeName(spec *volume.Spec) (string, error) {
+	csi, err := getCSISourceFromSpec(spec)
+	if err != nil {
+		glog.Error(log("plugin.GetVolumeName failed to extract volume source from spec: %v", err))
+		return "", err
+	}
+	if !isDriverNameValid(csi.Driver) {
```

Comment:
nit: This may not be the right place for this validation, since it's already in our system. Instead we should do this at the API level when accepting the `CSIVolumeSource` object (aka API validation).

Reply:
Agreed. I will put a TODO to move it in the api validation.